core/tests/test_utils.py

```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
+                    subprocess, 'Popen', mock_popen, expected_args=[
+                        (['python'],), (['python2'],)], expected_kwargs=[
+                            {'shell': True,}, {'shell': False}])
+                with popen_swap:
+                    function_that_invokes_popen()
+
+        Args:
+            obj: *. the Python object whose attribute you want to swap.
+            attr: str. The name of the function to be swapped.
+            newvalue: Function. The new function you want to use.
+            expected_args: None|list(tuple). The expected args that you
+                want this function to be invoked with. When its value is None,
+                args will not be checked.If the value type is list, the function
+                will check whether the called args is the first element in the
+                list. If matched, this tuple will be removed from the list.
+            expected_kwargs: None|list(dict). The expected key word args
+                you want this function to be invoked with. Similar to
+                expected_args.
+            called: bool. Whether the function is invoked. This will always be
+                checked.
+        """
+        original = getattr(obj, attr)
+        msg = 'Failed in check function %s.%s' % (obj.__name__, attr)
+        def wrapper(*args, **kwargs):
+            """Wrapper function for the new value. This function will do the
+            check before the wrapped function is invoked. After the function
+            finished, the wrapper will update how many times this function is
+            invoked.
+
+            Args:
+                args: tuple. The args passed into `attr` function.
+                kwargs: dict. The key word args passed into `attr` function.
+
+            Returns:
+                Result of `newvalue`.
+            """
+            wrapper.called = True
+            if expected_args is not None:
+                self.assertEqual(args, expected_args[0], msg=msg)
+                expected_args.pop(0)
+            if expected_kwargs is not None:
+                self.assertEqual(kwargs, expected_kwargs[0], msg=msg)
+                expected_kwargs.pop(0)
+            result = newvalue(*args, **kwargs)
+            return result
+        wrapper.called = False
+        setattr(obj, attr, wrapper)
+        error_occurred = False
+        try:
+            # This will show the detailed assert message.
+            self.longMessage = True
+            yield
+        except Exception as exception:
+            # Use this flag to indicate whether there was an error in the
+            # checks.
+            error_occurred = True
+            raise exception
+        finally:
+            self.longMessage = False
+            setattr(obj, attr, original)
+            # Only do the rest of the checks when there was no error.
+            # Otherwise error in this check will overwrite the error raised
```

Comment:
This was added to avoid the `called_times` error overwrite the other ones, as that functionality has been removed, I will aggregate it.

Reply:
Ah -- that actually raises another question. Suppose there are three elements in args, kwargs -- and the function is only called twice. Is that a problem? In any case we should document how that case is handled (and there should be unit tests too).

Reply:
This will be checked at the final part - whether the expected_args and expected_kwargs are empty.

> Could we make it such that the errors are aggregated? That would probably be more robust, rather than using "if" statements for fear of overwriting.

Also, I just realized it may be better to keep those separate. Otherwise, the error threw in the tested functions can not be tracked.